free5gc-operator/controllers/smfdeployment_controller.go

```diff
@@ -0,0 +1,648 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/go-logr/logr"
+	workloadv1alpha1 "github.com/nephio-project/api/nf_deployments/v1alpha1"
+)
+
+// SMFDeploymentReconciler reconciles a SMFDeployment object
+type SMFDeploymentReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+type SMFcfgStruct struct {
+	PFCP_IP string // N4
+	// N4_IP   string
+	// N11_IP  string
+}
+
+type SMFAnnotation struct {
+	Name      string `json:"name"`
+	Interface string `json:"interface"`
+	IP        string `json:"ip"`
+	Gateway   string `json:"gateway"`
+}
+
+// Check string -> int/string -uint16 conversion
+func getSMFResourceParams(smfSpec workloadv1alpha1.SMFDeploymentSpec) (int32, *apiv1.ResourceRequirements, error) {
+	// Placeholder for Capacity calculation. Reqiurce requirements houlw be calculated based on DL, UL.
+
+	// TODO: increase number of replicas based on NFDeployment.Capacity.MaxSessions
+	var replicas int32 = 1
+	var cpuLimit string
+	var cpuRequest string
+	var memoryLimit string
+	var memoryRequest string
+
+	// spec.Capacity.MaxSessions <1000 -> 128Mi
+	// spec.Capacity.MaxNFConnections <10 -> 128Mi
+	if smfSpec.Capacity.MaxSessions < 1000 && smfSpec.Capacity.MaxNFConnections < 10 {
+		cpuLimit = "100m"
+		cpuRequest = "100m"
+		memoryRequest = "128Mi"
+		memoryLimit = "128Mi"
+	} else {
+		cpuLimit = "500m"
+		cpuRequest = "500m"
+		memoryRequest = "512Mi"
+		memoryLimit = "512Mi"
+	}
+
+	resources := apiv1.ResourceRequirements{}
+	resources.Limits = make(apiv1.ResourceList)
+	resources.Limits[apiv1.ResourceCPU] = resource.MustParse(cpuLimit)
+	resources.Limits[apiv1.ResourceMemory] = resource.MustParse(memoryLimit)
+
+	resources.Requests = make(apiv1.ResourceList)
+	resources.Requests[apiv1.ResourceCPU] = resource.MustParse(cpuRequest)
+	resources.Requests[apiv1.ResourceMemory] = resource.MustParse(memoryRequest)
+
+	return replicas, &resources, nil
+}
+
+func constructSMFNadName(templateName string, suffix string) string {
+	return templateName + "-" + suffix
+	// return "n4"
+}
+
+// getSMFNad returns NAD label string composed based on the Nx interfaces configuration provided in SMFDeploymentSpec
+func getSMFNad(templateName string, spec *workloadv1alpha1.SMFDeploymentSpec) string {
+	var ret string
+
+	// n6CfgSlice := getIntConfigSlice(spec.Interfaces, "N6")
+	// n3CfgSlice := getIntConfigSlice(spec.Interfaces, "N3")
+	n4CfgSlice := getIntConfigSlice(spec.Interfaces, "n4")
+	// n9CfgSlice := getIntConfigSlice(spec.Interfaces, "N9")
+
+	ret = `[`
+	intfMap := map[string][]workloadv1alpha1.InterfaceConfig{
+		// "n3": n3CfgSlice,
+		"n4": n4CfgSlice,
+		// "n6": n6CfgSlice,
+		// "n9": n9CfgSlice,
+	}
+	// Need to sort inftMap by key otherwise unitTests might fail as order of intefaces in intfMap is not guaranteed
+	inftMapKeys := make([]string, 0, len(intfMap))
+	for interfaceName := range intfMap {
+		inftMapKeys = append(inftMapKeys, interfaceName)
+	}
+	sort.Strings(inftMapKeys)
+
+	noComma := true
```

Comment:
is this JSON we're producing here for something to consume? There's probably a more idiomatic way to do it that maps structs to JSON. That is, you create Go structs for the things you need, with JSON tags. In fact if these are Multus things they probably already have such types. So, you would import that module, construct the objects in Go, and then use json.Marshal. I *think* that would handle the sorting of the map keys too, but not sure.